### PR TITLE
[13.0][ADD] stock_picking_deliveryslip_subtotals

### PR DESCRIPTION
**Pull request by ChristianSantamaria**
Adds Sorted by Product ID and add a Subtotal section
Preview:
imagen

### CODE REVIEW

File: stock_picking_deliveryslip_subtotals/views/stock_picking_deliveryslip_subtotals.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        <xpath expr="//tr[@t-as='move_line']" position="inside">
+            <t t-set="cont" t-value="cont+1"/>
+            <t t-if="(len(o.move_line_ids.filtered(lambda x: x.product_id.id == move_line.move_id.product_id.id)) == cont)">
+                <tr>
+                    <td></td>
+                    <td class="border-top border-dark text-center">
+                    <strong style="color: #859b29;">Subtotal</strong>
+                    </td>
+                    <td class="border-top border-dark text-center">
+                    <span t-esc="sum(o.move_line_ids.filtered(lambda x: x.product_id.id == move_line.move_id.product_id.id).mapped('qty_done'))" t-options="{'widget': 'float', 'precision': decimal_precision}"/>
```
Comment (dalonsod) [inline stock_picking_deliveryslip_subtotals/views/stock_picking_deliveryslip_subtotals.xml]: This is inefficient (every line calculates aggregated data that could be obtained once). I'll take a look and try to provide you a better solution

**Review by dalonsod — CHANGES_REQUESTED** [review]
With the required change, locally tested OK. See comment

File: stock_picking_deliveryslip_subtotals/views/stock_picking_deliveryslip_subtotals.xml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+        <xpath expr="//tr[@t-as='move_line']" position="attributes">
+            <attribute name="t-foreach">o.move_line_ids.sorted(key=lambda x: x.product_id.id)</attribute>
+        </xpath>
+        
+        <xpath expr="//tr[@t-as='move_line']" position="inside">
+            <t t-set="cont" t-value="cont+1"/>
+            <t t-if="(lines_summary[move_line.product_id.id]['count'] == cont)">
+                <tr>
+                    <t t-if="has_serial_number">
```
Comment (dalonsod) [inline stock_picking_deliveryslip_subtotals/views/stock_picking_deliveryslip_subtotals.xml]: Please fix this:
```suggestion
                    <t t-if="has_serial_number" groups="stock.group_lot_on_delivery_slip">
```
With only `has_serial_number` will not work (picking lines could have lots, but disable their presentation by configuration)

**Review by dalonsod — APPROVED** [review]

**Comment by dalonsod** [comment]
@lmiguens-solvos could you review?

**Comment by lmiguens-solvos** [comment]
> @lmiguens-solvos could you review?

The idea is fine. All right.
But could the format of the subtotals be changed? For example, simply with a background? without the line ... (maybe it "mixes" a bit with the line below)
And about sorting ... by product id, wouldn't it be better by name?
This is my example:
image

**Comment by dalonsod** [comment]
> But could the format of the subtotals be changed? For example, simply with a background? without the line ... (maybe it "mixes" a bit with the line below)

OK

> And about sorting ... by product id, wouldn't it be better by name?

I agree... but @lromero-solvos initially said that current sorting should be preserved. Maybe we could ask him.

**Comment by ChristianSantamaria** [comment]
@lmiguens-solvos Done, they have removed the top border and added a `background-color: whitesmoke`. What do you think?
imagen

**Comment by lmiguens-solvos** [comment]
> @lmiguens-solvos Done, they have removed the top border and added a `background-color: whitesmoke`. What do you think?
> imagen

Perfect! I like it much better

**Comment by lmiguens-solvos** [comment]
> I agree... but @lromero-solvos initially said that current sorting should be preserved. Maybe we could ask him.

Ok, I thought there was no order and this was chosen. But if there already was, for now we are not going to change it. The customer will think (or not).

**Review by lmiguens-solvos — APPROVED** [review]
LGTM

**Comment by dalonsod** [comment]
> Ok, I thought there was no order and this was chosen. But if there already was, for now we are not going to change it. The customer will think (or not).

That's OK, no problem changing it if needed later.
@ChristianSantamaria please squash commits

**Comment by lromero-solvos** [comment]
Hello, everybody

I had indicated that this delivery note had to be ordered as another report. Now I don't remember which one. Now, let it be...

**Pull request merged**